src/tactic/solve_by_elim.lean

```diff
@@ -39,42 +57,73 @@ do (hs, gex, hex, all_hyps) ← decode_simp_arg_list hs,
    else return hs
 
 /--
-The internal implementation of `solve_by_elim`, with a limiting counter.
+Configuration options for `solve_by_elim`.
+
+* `accept : list expr → tactic unit` determines whether the current branch should be explored.
+   It is passed the original goals as a `list expr` argument
```

Comment:
Sorry, this line is still confusing me. I would guess first that it's looking for a list of `mvar`s, and second that it's looking for a list of the types of these `mvar`s. In the code, it looks like you give it neither -- you instantiate the `mvar`s before passing them to `accept`. Is that optional, or required? What's structure is this `list expr` argument expected to have?

Reply:
Okay, I've tried again.

The idea is that when `solve_by_elim` is first invoked, we call `get_goals`, and stash them. Now, at every step in the search tree, we take those "original goals", instantiate metavariables in them (as those original goals have by now been partly solved by previous `apply` steps), and pass them to the `accept` tactic. Thus the `accept` tactic can inspect the history of `apply` steps that have succeeded up to this point, by looking at how the original metavariables from `get_goals` have been partially unified.

The main change I made in the docs is to replace phrases like "original goals", explicitly saying that `solve_by_elim` calls `get_goals`:

> It is passed the original metavariables reported by `get_goals` when `solve_by_elim` started, as a `list expr` argument (these metavariables may by now have been partially solved by previous `apply` steps), ...

I agree this is non-obvious, and my doc-strings were not adequate, so please tell me if we're still not there yet!

Reply:
`accept` is run *before* the current goals are explored, right?  It would be good to clarify that in the docstring.

Reply:
Clarified!